botocore/credentials.py

```diff
@@ -436,11 +444,169 @@ def get_frozen_credentials(self):
+class AssumeRoleCredentialFetcher(object):
+    def __init__(self, client_creator, source_credentials, role_arn,
+                 extra_args=None, mfa_prompter=None, cache=None,
+                 expiry_window_seconds=60 * 15):
+        """
+        :type client_creator: callable
+        :param client_creator: A callable that creates a client taking
+            arguments like ``Session.create_client``.
+
+        :type source_credentials: Credentials
+        :param source_credentials: The credentials to use to create the
+            client for the call to AssumeRole.
+
+        :type role_arn: str
+        :param role_arn: The ARN of the role to be assumed.
+
+        :type extra_args: dict
+        :param extra_args: Any additional arguments to add to the assume
+            role request using the format of the botocore operation.
+            Possible keys include, but may not be limited to,
+            DurationSeconds, Policy, SerialNumber, ExternalId and
+            RoleSessionName.
+
+        :type mfa_prompter: callable
+        :param mfa_prompter: A callable that returns input provided by the
+            user (i.e raw_input, getpass.getpass, etc.).
+
+        :type cache: dict
+        :param cache: An object that supports ``__getitem__``,
+            ``__setitem__``, and ``__contains__``.  An example of this is
+            the ``JSONFileCache`` class in aws-cli.
+
+        :type expiry_window_seconds: int
+        :param expiry_window_seconds: The amount of time, in seconds,
+        """
+        self._client_creator = client_creator
+        self._source_credentials = source_credentials
+        self._role_arn = role_arn
+        self._expiry_window_seconds = expiry_window_seconds
+
+        if extra_args is None:
+            self._assume_kwargs = {}
+        else:
+            self._assume_kwargs = deepcopy(extra_args)
+        self._assume_kwargs['RoleArn'] = self._role_arn
+
+        self._role_session_name = self._assume_kwargs.get('RoleSessionName')
+        self._using_default_session_name = False
+        if not self._role_session_name:
+            self._role_session_name = 'botocore-session-%s' % (
+                int(time.time()))
+            self._assume_kwargs['RoleSessionName'] = self._role_session_name
+            self._using_default_session_name = True
+
+        self._mfa_prompter = mfa_prompter
+        if self._mfa_prompter is None:
+            self._mfa_prompter = getpass.getpass
+
+        self._cache = cache
+        if self._cache is None:
+            self._cache = {}
+
+        self._cache_key = self._create_cache_key()
+
+    def _create_cache_key(self):
+        """Create a predictable cache key for the current configuration.
+
+        The cache key is intended to be compatible with file names.
+        """
+        # On windows, ':' is not allowed in file names, so we'll
+        # replace them with '_' instead.
+        cache_key = self._role_arn.replace(':', '_').replace(os.path.sep, '_')
```

Comment:
Interesting.  The cache key will no longer have the profile name as part of its cache key which means you will share cached credentials across different profiles if they happen to assume the same role.  e.g

```
[profile one]
role_arn=MYROLE
[profile two]
role_arn=MYROLE
```

This isn't a safe assumption because different profiles could have different arguments but still assume the same role (external_id, mfa_serial, etc).

Reply:
That's a good point. How about doing a hash of all of `self._assume_role_kwargs`? It would also serve to make the generation much simpler.